service/src/main/java/crawlercommons/urlfrontier/service/AbstractFrontierService.java

```diff
@@ -872,4 +873,94 @@ public void close() throws IOException {
     public abstract void getURLStatus(
             crawlercommons.urlfrontier.Urlfrontier.URLStatusRequest request,
             io.grpc.stub.StreamObserver<URLItem> responseObserver);
+
+    public void listURLs(
+            crawlercommons.urlfrontier.Urlfrontier.ListUrlParams request,
+            io.grpc.stub.StreamObserver<crawlercommons.urlfrontier.Urlfrontier.URLItem>
+                    responseObserver) {
+
+        long maxURLs = request.getSize();
+        long start = request.getStart();
+        String key = request.getKey();
+
+        final String normalisedCrawlID = CrawlID.normaliseCrawlID(request.getCrawlID());
+
+        // 100 by default
+        if (maxURLs == 0) {
+            maxURLs = 100;
+        }
+
+        LOG.info(
+                "Received request to list URLs [size {}; start {}; crawlId {}, key {}]",
+                maxURLs,
+                start,
+                normalisedCrawlID,
+                key);
+
+        long totalCount = -1;
+        long sentCount = 0;
+
+        synchronized (getQueues()) {
+            Iterator<Entry<QueueWithinCrawl, QueueInterface>> qiterator =
+                    getQueues().entrySet().iterator();
+
+            while (qiterator.hasNext() && sentCount < maxURLs) {
+                Entry<QueueWithinCrawl, QueueInterface> e = qiterator.next();
+
+                // check that it is within the right crawlID
+                if (!e.getKey().getCrawlid().equals(normalisedCrawlID)) {
+                    continue;
+                }
+
+                // check that it is within the right key/queue
+                if (key != null && !key.isEmpty() && !e.getKey().getQueue().equals(key)) {
+                    continue;
+                }
+
+                Iterator<URLItem> urliter = urlIterator(e);
+
+                while (urliter.hasNext()) {
+                    totalCount++;
+                    if (totalCount < start) {
+                        urliter.next();
+                    } else if (sentCount < maxURLs) {
+                        responseObserver.onNext(urliter.next());
+                        sentCount++;
+                    } else {
+                        break;
+                    }
+                }
+            }
+        }
+
+        responseObserver.onCompleted();
+    }
+
+    protected abstract Iterator<URLItem> urlIterator(
```

Comment:
This is only used once I think. Can't we simply have the one below with 0, and Integer.maxInt as values when called?

Reply:
Yes, I removed the redundant iterator constructors